File: widgets/iconbutton.py
import qtawesome as qta

from typing import Union

from PyQt5.QtCore import Qt, QSize
from PyQt5.QtGui import QIcon
from PyQt5.QtWidgets import QHBoxLayout, QLabel, QPushButton


class IconButton(QPushButton):
    def __init__(self, icon: Union[str, QIcon], text: str):
        super(IconButton, self).__init__()

        # credit: https://stackoverflow.com/a/61130602
        self.setIcon(icon)
        self.setIconSize(QSize(64, 64))
        # align the icon left
        self.setStyleSheet("text-align: left;")
        self.setLayout(QHBoxLayout())

        self.label = QLabel(text)
        self.label.setAlignment(Qt.AlignCenter | Qt.AlignVCenter)
        self.label.setAttribute(Qt.WA_TransparentForMouseEvents, True)
        # TODO: find better way to show all text
        self.label.setStyleSheet("margin-top: 100px; margin-bottom: 100px;")

        self.layout().addWidget(self.label)

    def setIcon(self, icon: Union[str, QIcon]):
        if isinstance(icon, str):
            icon = qta.icon(icon)

        super(IconButton, self).setIcon(icon)

File: src/iconbutton.cpp
#include "iconbutton.h"

#include <QHBoxLayout>
#include <QVariantMap>

#include "main.h"

IconButton::IconButton(QWidget* parent)
  : QPushButton{ parent }
  , mAnimation{ this }
{
  setIconSize(QSize(64, 64));

  setStyleSheet("IconButton { border: 0px; padding: 50px; text-align: left; "
                "background: rgba(0,0,0,70%); color: rgba(220, 220, 220, "
                "100%); } IconButton:!enabled { background: rgba(0,0,0,30%); "
                "color: rgba(122, 122, 122, 100%); }");

  mLabel.setAlignment(Qt::AlignCenter | Qt::AlignVCenter);
  mLabel.setStyleSheet(
    "QLabel { background: transparent; color: rgba(220, 220, 220, 100%); } "
    "QLabel:!enabled { color: rgba(122, 122, 122, 100%); }");
  mLabel.setAttribute(Qt::WA_TransparentForMouseEvents, true);

  QHBoxLayout* layout = new QHBoxLayout();
  layout->addWidget(&mLabel);
  setLayout(layout);
}

void
IconButton::setIcon(const QIcon& icon)
{
  QPushButton::setIcon(icon);
  QPushButton::setText("");
}

void
IconButton::setIcon(int charCode)
{
  QPushButton::setIcon(QIcon());
  QPushButton::setText(QString::fromStdU32String(
    std::u32string({ static_cast<char32_t>(charCode) })));
}

void
IconButton::setIcon(const QString& iconString)
{
  QPushButton::setIcon(QIcon());
  QPushButton::setText(iconString);
}

void
IconButton::setText(const QString& text)
{
  mLabel.setText(text);
}

void
IconButton::setSpinner()
{
    QVariantMap options;
    options.insert("anim", QVariant::fromValue(&mAnimation));
    setIcon(awesome()->icon(fa::fa_solid, fa::fa_spinner, options));
    QPushButton::setText("");
}
